### ISSUE TITLE
size returns possibly incorrect value for { length: 3 }

### ISSUE THREAD

**Issue by jacobp100** [lodash/lodash#2683]
If you do `size({ length: 3 })`, it returns `3`. I do sort of get the logic behind this, but in this case, the object was not array like. [Example](https://github.com/jacobp100/recora/commit/9e708a861cae5f0fff4f5872fd9e032eb60d74a1#diff-dfedccebf267017bb1670428731d926dR52)

Either this should be documented, or the logic might need changing.

**Comment by jdalton** [comment]
Hi @jacobp100!

It is correct and [documented](https://lodash.com/docs#size). Here's the relevant bit:
> Gets the size of `collection` by returning its length for array-like values

**Comment by emmettnicholas** [comment]
I realize that every "`length`" issue is closed with the same explanation (https://github.com/lodash/lodash/issues/1870, https://github.com/lodash/lodash/issues/1966, https://github.com/lodash/lodash/issues/2352, https://github.com/lodash/lodash/issues/2338, https://github.com/lodash/lodash/issues/2683, etc.), but I'm curious to know whether "because it's array-like" is merely an explanation/documentation of broken behavior, or whether this is actually considered desirable behavior.

In other words, in the next major of Lodash, is this behavior a candidate for "fixing", or is it safe to assume that Lodash will continue to special-case `length` forever?

**Comment by jdalton** [comment]
@emmettnicholas It's the desired behavior. The "Collections" methods prioritize array-like values over objects.

**Comment by emmettnicholas** [comment]
Thanks for the clarification.

IMO the advantage of being able to use "collections" methods on `arguments`/`NodeList` isn't worth the trade-off of _not_ being able to use "collections" methods on objects whose keys are not pre-determined, e.g. come from user input. In my particular use case, I'm building a user-facing SDK whereby the user supplies objects with arbitrary key/value pairs. Currently Lodash is unusable without a a layer to sanitize all "length" keys first.

**Comment by jdalton** [comment]
Several of the collections methods have object specific alternatives. Beyond that you do what you need to do. Lodash and Underscore before it has had collections since 2009 and it seems to be working out OK.

**Comment by emmettnicholas** [comment]
> It seems to be working out OK.

Sure, but that shouldn't preclude Lodash from improving in future major releases, especially if certain issues remain consistent sources of confusion.

- Since the time that Underscore introduced collection methods in 2009, json has become an even more established data interchange format. This `length` issue means lodash definitely cannot be used to operate on arbitrary json data.
- The necessity to handle `arguments` is going away as rest params [gain support](http://eslint.org/docs/rules/prefer-rest-params).
- Other array-like structures like `NodeList` are uncommon enough that they could be handled with a special convert-to-array method.
